### PR TITLE
Fix media merge with variants and compound variants

### LINKED ISSUE #34: Css function Composer type extends String type require length property is not needed
This problem is almost same as https://github.com/Temzasse/stitches-native/pull/26.
The css function needs same patch.
スクリーンショット 2022-10-31 15 25 45

### PR DESCRIPTION
**Pull request by tkow**
Closes #34
Check after #31 , #32.

I fixed media properties can be used with variants and compound variants styles.

I removed [these code](https://github.com/Temzasse/stitches-native/compare/main...MOCHI-inc-JAPAN:stitches-native:fix/media-merge-with-variants?expand=1#diff-55390e2d0a61588763c772b215b79bd7259d702c4eb80e315506b87f1fa5c8d2L92-L111), because it's been ever ignored and not worked due to `key` format is used  instead of `@key`. In addition, now media properties must be kept as media variants style for work. 

I add new examples of this features.
スクリーンショット 2022-10-29 19 05 25

### CODE REVIEW

**Comment by Temzasse** [comment]
Thank you for these PRs! On a quick glance they look very good 👍🏻 I'll try to find some time to review and test them locally 🙂 

It looks like these three PRs build on top of each other so I'm wondering if it would be easier for me to review just the latest PR that contains all the fixes 🤔 They are all related to responsive styles so it wouldn't be too bad to review all the fixes as one big PR.

Is it okay to you if I just review and test this PR and close the other two?

**Comment by tkow** [comment]
@Temzasse 

> Is it okay to you if I just review and test this PR and close the other two?

Yes, of course.
Thank you for reaction and taking your time.

**Comment by Temzasse** [comment]
Thank you for setting up the Expo Web support in the example app! 🙌🏻 

I noticed that running the example app in iOS Simulator doesn't seem to be working anymore because it's not able to resolve `stitches-native`. I think it's very nice to be able to test the lib in a web browser but the main target of this lib is native app development so it's very important to be able to test the app inside iOS/Android simulator.

I wonder why the web version is working but not the native one 🤔

**Comment by tkow** [comment]
Sorry for this pitfall. It may be caused by expo and react-native latest versions.
I will check  tomorrow both android and iOS simulator. Let me share if I find the reason.

**Comment by Temzasse** [comment]
I think we can have both metro and webpack configs in the example Expo app. Metro will be used for native and webpack for web. I took a look at this RN library called [Zeego](https://github.com/nandorojo/zeego/tree/master/examples/expo) that has both config files in the example app. I will play around with the bundler setup today and see if I can get it to work.

**Comment by tkow** [comment]
Ah... I see. I certainly removed metro.config.js. The reason is  mentioned in https://github.com/Temzasse/stitches-native/pull/31. But, you're right, we must use metro or repack as  native module bundler. Especially I don't  know whether we can use repack with expo so we should use metro. Thank you for investigation and I'll try to fix it too, though no problem as if you finish to work earlier than me.

**Comment by Temzasse** [comment]
I think I got it working with both webpack and metro. Can you give me [rights](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) to push to your PR branch?

File: example/package.json
Comment (Temzasse) [inline example/package.json]: Move this to dev deps.

**Comment by tkow** [comment]
@Temzasse Thank you, I worked just restoring metro.config.js, too. If you do another way and want more fix, could you do after merging my code?

File: example/webpack.config.js
Comment (Temzasse) [inline example/webpack.config.js]: I think we can add `react-native` and `react-native-dom` aliases here as well like in [Zeego](https://github.com/nandorojo/zeego/blob/master/examples/expo/webpack.config.js#L21):

```suggestion
  Object.assign(config.resolve.alias, {
    react: path.join(__dirname, 'node_modules', 'react'),
    'react-native': path.join(__dirname, 'node_modules', 'react-native-web'),
    'react-native-web': path.join(__dirname, 'node_modules', 'react-native-web'), // prettier-ignore
  });
```

**Review by Temzasse — CHANGES_REQUESTED** [review]
This the first review round for the example app related changes.

Next I will take a look at the lib changes.

**Comment by tkow** [comment]
スクリーンショット 2022-10-31 0 12 19
I tested and worked in our local environment like this.

**Comment by Temzasse** [comment]
> @Temzasse Thank you, I worked just restore metro.config.js, too. If you do another way and want more fix, could you do after merging my code?

Sure that works too 👍🏻

File: src/internals/utils.js
Comment (Temzasse) [inline src/internals/utils.js]: There is no need for priority ordering for these keys - they can be in any order. This means that you can derive them from the `THEME_VALUES` object from `constants.ts`:

```js
const THEME_KEYS = Object.keys(THEME_VALUES);
```

**Comment by tkow** [comment]
@Temzasse 
Sorry, let you review many times.
If you still have problems to work, I will set permission my repository.

File: src/internals/utils.js
Comment (Temzasse) [inline src/internals/utils.js]: NIT: typo `popeties` -> `properties`

File: src/internals/utils.js
Comment (Temzasse) [inline src/internals/utils.js]: NIT: a better name for this function would be something like `getThemeScaleKey` and instead of using variable called `configKey` in various places in this file a more accurate name would be `scaleKey`.

File: src/internals/utils.js
Comment (Temzasse) [inline src/internals/utils.js:167]: Could you add a comment for this case so that it's easier to understand what is happening here when I or someone else comes back to this codebase after some time 😄

File: src/internals/index.js
Comment (Temzasse) [inline src/internals/index.js]: This if could be merged with the below if to reduce the number of ifs:

```js
if (breakpoint && propValue[breakpoint] !== undefined) {}
```

File: src/internals/index.js
Comment (Temzasse) [inline src/internals/index.js:161]: This can also be simplified so that there is only one if:

```js
if (val === true || typeof val === 'string') {
  styleSheetKey = `${prop}_${propValue[breakpoint]}`;
}
```

**Review by Temzasse — CHANGES_REQUESTED** [review]
All in all changes to the internals look very good 👍🏻 

Found a few places where we can cleanup the code a bit.

I will still test the implementation a bit more in order to make sure all different cases of media styles work as expected.

**Comment by Temzasse** [comment]
@tkow Thank you for being patient with my many reviews 🙏🏻  After making the changes requested in my last review round I can merge this PR and fix anything else that comes up on my own 🙂

**Comment by tkow** [comment]
@Temzasse 
Sure. Thank you for many advices.  I can take time an hour now, fix them at least in an hour. 

By the way, I don't know why, but [this](https://github.com/Temzasse/stitches-native/pull/33#issuecomment-1296281138) instruction can't be applied.(UI doesn't exist)

スクリーンショット 2022-10-31 0 52 00

 So, in already our repository public you may be able to push without any setting. If you can't do so and need to do, I'll suggest another way after fixing the your review points.

File: src/internals/utils.js
Comment (tkow) [reply src/internals/utils.js]: Okey, but colors key is also included. So, I'll rename getThemeKey and themeKey.

File: src/internals/utils.js
Comment (tkow) [reply src/internals/utils.js:167]: DONE

**Review by Temzasse — APPROVED** [review]
Awesome job! Thank you very much for this 🙌🏻 ✨ 

I will merge this after I have had the time to test it a bit more 🙂

**Comment by tkow** [comment]
Sorry, after review approved, but I found the type problem for https://github.com/Temzasse/stitches-native/issues/34 and, example's @type/react version's not compatible with installed react. Then, I add compoundVariants example for check.
These are not so much change and essentially bug fix, thus I'll merge it to PR.

See,

https://github.com/Temzasse/stitches-native/pull/33/commits/61d83b8e816d82f37f76fe724f2df053bf682144

and

スクリーンショット 2022-10-31 16 22 29

**Comment by tkow** [comment]
@Temzasse 

~In addition, I don't know why but found the bug defaultVariants combination doesn't match compoundVariants, though 
the  passed props  work. If you want to check, please change my example.~

```typescript
defaultVariants: {
      heading: 'h1',
      underlined: true,
},
```

~This bug may have ever been and not very serious because passing props explicitly  so, could I ask you to get it out of scope in this PR?~
I can fix it easily, so bring it to this PR. Now example looks that.

スクリーンショット 2022-10-31 17 52 15

As I'm already using these new features as private git submodules. So, when I find something wrong and it's mission critical for our development, I may work for this library.

Stitches and it for the native, really  most awesome library as I have ever seen especially in the point of cross platform availability while it may be hard with streaming SSR , and I want to keep using them for long. I appreciate your nice work public.

**Comment by Temzasse** [comment]
@tkow thank you so much for these fixes! 👏🏻 

Let's not add any more fixes outside of the media styles related bug fixes or improvements into this PR so that that I'm able to review things more efficiently.

I will try to merge this as soon as possible so that you can get these fixes from the npm instead of your git submodule 🙂

**Review by Temzasse — APPROVED** [review]

**Pull request merged**
